File: HappyNumber/HappyNumber.ts
/*
 *  Some numbers are happy. Well they're only happy if sum of squares
 *  of their digits ends up as 1!
 */

function isHappy(n: number): boolean {
	if (n == 1 || n == 7) return true;

	let sum: number = n;

  	while (sum > 9) {
  		sum = sum
      		.toString()
      		.split('')
      		.map((digit) => {
      			return Number(digit);
      		})
      		.reduce((partialSum, a) => Math.pow(a, 2) + partialSum, 0);

    	if (sum == 1 || sum == 7) {
      		return true;
    	}
	}

  	return false;
}

File: HappyNumber/HappyNumber.py
#isHappyNumber() will determine whether a number is happy or not    
def isHappyNumber(num):    
    rem = sum = 0;    
        
    #Calculates the sum of squares of digits    
    while(num > 0):    
        rem = num%10;    
        sum = sum + (rem*rem);    
        num = num//10;    
    return sum;    
        
num = 82;    
result = num;    
     
while(result != 1 and result != 4):    
    result = isHappyNumber(result);    
     
#Happy number always ends with 1    
if(result == 1):    
    print(str(num) + " is a happy number");    
#Unhappy number ends in a cycle of repeating numbers which contain 4    
elif(result == 4):    
    print(str(num) + " is not a happy number");   